Improve chat and additional small fixes for party interaction

Ellie42 opened the pull request:
This PR focuses on adding support for local/global/party/private chat messages, and only sending the messages where they need to go.

Party and private chat messages seem to be essentially the same structure with a list of target character IDs and a flag to say if it's private or party. These target IDs need to be stripped before the client will accept the message.

### Additional
* Adds a notify message whenever a user calls SET_USER_BINARY which seems to fix a few missing state changes, and also covers a user's initial entry into the stage.
* Sends the correct user update messages when a player logs out or closes their client
* Add optional extra logging for outbound messages
* Add a `CharID` line to message dump logs to show who is sending or receiving the message

Let me know if I've got the right idea with this PR, with code style/naming/anything.

This has been tested in town with 2 and 3 players, in a quest with 2.

Andoryuuta commented:
Hi, thanks for the PR!

I've only skimmed over the changes, but they look great from what I can tell. I'll look more closely at when I get home later and try to get it merged. :)

Removing the `MsgSysNotifyUserBinary` in the `doStageTransfer` function might be an issue though, is a newly joining client properly notified of the existing stage's characters correctly?

Ellie42 commented:
The entree client is still sent the notifications of all other clients in the loop below, I have removed the entree client notifying existing clients. 
This is because another change I added is just sending `MSG_SYS_NOTIFY_USER_BINARY` on every `MSG_SYS_SET_USER_BINARY`, which seems to cover a few extra cases around player's party join/create/ready state too.
When a player joins the stage, IIRC it sends a few `MSG_SYS_SET_USER_BINARY` messages, which in turn notifies all existing members. I can more thoroughly confirm this later though as I'm working atm.

The player joining stage state appears to be working in the same way as the master branch from what I can tell.

Andoryuuta commented:
Ah, sorry! I wasn't paying attention this morning, I had thought you had removed the notifications to the entree client.

I'll merge it now. Thanks!
